content/ytt/docs/latest/lang-ref-ytt-schema.md

```diff
@@ -0,0 +1,244 @@
+
+### Inferring Defaults for Mappings
+
+The set of items of a map are its default: any item missing will be automatically inserted with _its_ defaults.
+
+- if `load_balancer` is omitted from supplied Data Values, it is inserted:
+    - `load_balancer.enabled` is `true`
+    - `load_balancer.static_ip` is `""`
+- if `load_balancer` is specified with only `static_ip`, `enabled` is `true`.
+
+When the Configuration Consumer supplies an empty map (i.e. `{}`), all items are defaulted:
+
+For example,
+
+    ```yaml
+    load_balancer: {} 
+    ```
+
+is defaulted to:
+
+```yaml
+load_balancer:
+  enable: true
+  static_ip: ""
+```
+
+### Inferring Defaults for Arrays
+
+The default value for all arrays is an empty array. Arrays are the only type for which defaults are **not** inferred.
+
+From the example, [above](#inferring-types),
+
+- `app_domains`, if not otherwise specified is `[]`
+- `databases`, if not otherwise specified is `[]`
+
+When values for an array _are_ provided by the Configuration Consumer, each item is defaulted based on the type specified in the schema.
+
+Focusing on just `databases`, if the Configuration Consumer supplies this Data Value:
+
+```yaml
+#@data/values
+---
+databases:
+- name: uaa
+- name: capi
+  host: capi-db.svc.cluster.local
+  secretRef:
+    name: capi-db-credentials
+- {}
+```
+
+The resulting Data Values (i.e. that are supplied to templates), would be:
+
+```yaml
+databases:
+- name: uaa
+  adapter: postgresql
+  host: ""
+  port: 5432
+  user: admin
+  secretRef:
+    name: ""
+- name: capi
+  adapter: postgresql
+  host: capi-db.svc.cluster.local
+  port: 5432
+  user: admin
+  secretRef:
+    name: capi-db-credentials
+- name: ""
+  adapter: postgresql
+  host: ""
+  port: 5432
+  user: admin
+  secretRef:
+    name: ""
+```
+
+## Defining Schema Explicitly
+
+Configuration Authors may override inferred typing and defaults through annotations.
+
+### @schema/type
+
+Explicitly configures the type of the annotated node. Currently, the only supported configuration is whether to allow any type or not.
+
+```yaml
+@schema/type any=True
+```
+
+- `any` (`bool`) — whether or not any and all types are permitted on this node (and its children)
+    - to avoid confusion, this keyword argument is mutually exclusive to all other arguments; if this argument is
+      provided along with others, an error results.
+__
+
+### @schema/nullable
+
+Extends the type of the node to include "null" _and_ to be "null" by default.
+
+```yaml
+@schema/nullable
+```
+
+_Example: Nullable map and string_ 
+
+```yaml
+#@schema/match data_values=True
+---
+#@schema/nullable
+aws:
+  username: admin
+  password: "1234"
+
+#@schema/nullable
+name: dev
+```
+
+If evaluated without any data values, it results in `aws` and `name` null by default.
+```yaml
+aws: null
+name: null
+```
+
+If evaluated with the following data values:
+
+```yaml
+#@data/values
+---
+aws:
+  username: sa
+
+name: acceptance
```

Comment:
What do you think of keeping `name` as `null` for this example... to illustrate how `null` _does_ flow through, if not specified... ?

```suggestion
#@data/values
---
aws:
  username: sa
```

Reply:
good idea!